src/transaction/mod.rs

```diff
@@ -81,24 +114,61 @@ impl Transaction {
     /// to determine which segment to write to. It calls the corresponding transactional event segment
     /// writer to write the data to segmentstore server.
-    pub async fn write_event(&mut self, key: Option<String>, event: Vec<u8>) -> Result<(), TransactionError> {
+    pub async fn write_event(
+        &mut self,
+        routing_key: Option<String>,
+        event: Vec<u8>,
+    ) -> Result<(), TransactionError> {
         self.error_if_closed()?;
 
-        let segment = if let Some(key) = key {
-            self.segments.get_segment_for_string(&key)
+        let size = event.len();
+        let (tx, rx) = oneshot::channel();
+        if let Some(pending_event) = PendingEvent::with_header(routing_key, event, tx) {
+            let append_event = Incoming::AppendEvent(pending_event);
+            if let Err(e) = self.sender.send((append_event, size)).await {
+                error!(
+                    "failed to write to transaction {:?} due to {:?}",
+                    self.info.txn_id, e
+                );
+                return Err(TransactionError::TxnSegmentWriterError {
+                    error_msg: format!("{:?}", e),
+                });
+            }
+            self.event_handles.push(rx);
         } else {
-            self.segments.get_segment(get_random_f64())
-        };
-        let transaction = self
-            .inner
-            .get_mut(&segment)
-            .expect("must get segment from transaction");
-        let span = info_span!("Transaction", transactionId = %self.info.txn_id);
-        transaction
-            .write_event(event, &self.factory)
-            .instrument(span)
-            .await
-            .context(TxnSegmentWriterError {})?;
+            let e = rx.await.expect("get error");
+            error!(
+                "failed to write to transaction {:?} due to {:?}",
+                self.info.txn_id, e
+            );
+            return Err(TransactionError::TxnSegmentWriterError {
+                error_msg: format!("{:?}", e),
+            });
+        }
+
+        let mut ack_up_to = 0;
+        for event in &mut self.event_handles {
+            match event.try_recv() {
+                Ok(res) => {
+                    if let Err(e) = res {
+                        return Err(TransactionError::TxnSegmentWriterError {
+                            error_msg: format!("error when writing event: {:?}", e),
+                        });
+                    } else {
+                        ack_up_to += 1;
+                    }
+                }
+                Err(TryRecvError::Empty) => {
+                    break;
+                }
+                Err(TryRecvError::Closed) => {
+                    return Err(TransactionError::TxnSegmentWriterError {
+                        error_msg: "event handle closed, cannot get result for event".to_string(),
+                    });
+                }
+            }
+        }
+        self.event_handles.drain(0..ack_up_to);
```

Comment:
Nit (feel free to ignore): This could probably be done more cleanly with `peek` and `pop`.